### PR TITLE
Fix `ContainerScreenEvent$Render$Background` running after renderables being drawn.

### LINKED ISSUE #608: `ContainerScreenEvent$Render$Background` runs after renderables are drawn
**Minecraft Version:** 1.20.4

**NeoForge Version:** 20.4.42

**Description of issue:**
In 1.20.1 the order in `AbstractContainerScreen::render` used to be:
- `AbstractContainerScreen::renderBg`
  - `Screen::renderBackground`
    - Draws the black transparent background overlay
  - Draws the menu background sprite
- `ContainerScreenEvent$Render$Background` (NeoForge event)
- `Screen::render` (super method)
  - Draws the screen renderables
- ...

In 1.20.2+ `AbstractContainerScreen::renderBg` has been removed in favor of using `Screen::renderBackground` which is now implemented in subclasses. Therefore the order in `AbstractContainerScreen::render` has changed to:
- `Screen::render` (super method)
  - `Screen::renderBackground`
    - Draws the black transparent background overlay
    - Draws the menu background sprite
  - Draws the screen renderables
- `ContainerScreenEvent$Render$Background` (NeoForge event)
- ...

This new behavior makes it impossible to render background sprites which are drawn behind renderables like buttons & widgets. 
For intentionally rendering sprites in front of renderables `ContainerScreenEvent$Render$Foreground` already exists.

My suggestion for restoring the 1.20.1 behavior would be to move firing `ContainerScreenEvent$Render$Background` to `Screen::render` (with an `instanceof` check for `AbstractContainerScreen`) right in-between the call to `Screen::renderBackground` and screen renderables being drawn.

Moving `ContainerScreenEvent$Render$Background` to `AbstractContainerScreen::renderBackground` is not a good option as it would fire after the black transparent background overlay is drawn, but before the menu background sprite is.

### PR DESCRIPTION
**Pull request by Spinoscythe**
Fixes #608

### CODE REVIEW

**Comment by XFactHD** [comment]
This is not the correct fix in my opinion. Instead of moving the event call to `Screen` behind an `instanceof` check, the event call should be moved to `AbstractContainerScreen#renderBackground()` after the `AbstractContainerScreen#renderBg()` call.

**Comment by Spinoscythe** [comment]
> This is not the correct fix in my opinion. Instead of moving the event call to `Screen` behind an `instanceof` check, the event call should be moved to `AbstractContainerScreen#renderBackground()` after the `AbstractContainerScreen#renderBg()` call.

fixed!

File: patches/net/minecraft/client/gui/screens/Screen.java.patch [hunk 1]
```diff
@@ -1,5 +1,13 @@
 --- a/net/minecraft/client/gui/screens/Screen.java
 +++ b/net/minecraft/client/gui/screens/Screen.java
+@@ -113,7 +_,6 @@
+     @Override
+     public void render(GuiGraphics p_281549_, int p_281550_, int p_282878_, float p_282465_) {
+         this.renderBackground(p_281549_, p_281550_, p_282878_, p_282465_);
+-
```
Comment (XFactHD) [inline patches/net/minecraft/client/gui/screens/Screen.java.patch]: Rogue patch, please revert this

File: patches/net/minecraft/client/gui/screens/inventory/AbstractContainerScreen.java.patch [hunk 2]
```diff
@@ -23,7 +15,12 @@
          ItemStack itemstack = this.draggingItem.isEmpty() ? this.menu.getCarried() : this.draggingItem;
          if (!itemstack.isEmpty()) {
              int l1 = 8;
-@@ -156,13 +_,16 @@
+@@ -152,17 +_,21 @@
+     @Override
+     public void renderBackground(GuiGraphics p_295206_, int p_295457_, int p_294596_, float p_296351_) {
+         super.renderBackground(p_295206_, p_295457_, p_294596_, p_296351_);
++        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ContainerScreenEvent.Render.Background(this, p_295206_, p_295457_, p_295457_));
```
Comment (XFactHD) [inline patches/net/minecraft/client/gui/screens/inventory/AbstractContainerScreen.java.patch]: This is still the wrong place, it should be after the `renderBg()` call

File: src/main/java/net/neoforged/neoforge/client/event/ContainerScreenEvent.java [hunk 3]
```diff
@@ -8,6 +8,7 @@
 import net.minecraft.client.gui.GuiGraphics;
 import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
 import net.neoforged.bus.api.Event;
+import net.neoforged.bus.api.ICancellableEvent;
```
Comment (XFactHD) [inline src/main/java/net/neoforged/neoforge/client/event/ContainerScreenEvent.java]: Pointless change, please revert this

**Review by XFactHD — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/client/event/ContainerScreenEvent.java (hunk 3 above)
Comment (Spinoscythe) [reply src/main/java/net/neoforged/neoforge/client/event/ContainerScreenEvent.java]: it's just a javadoc fix. I don't think we need to go head over heels for this.

Comment (Matyrobbrt) [reply src/main/java/net/neoforged/neoforge/client/event/ContainerScreenEvent.java]: Random unrelated fixes really do not help with tracking the history of changes in a file.

Comment (Spinoscythe) [reply src/main/java/net/neoforged/neoforge/client/event/ContainerScreenEvent.java]: ok

File: patches/net/minecraft/client/gui/screens/inventory/AbstractContainerScreen.java.patch (hunk 2 above)
Comment (Spinoscythe) [reply patches/net/minecraft/client/gui/screens/inventory/AbstractContainerScreen.java.patch]: fixed!

File: patches/net/minecraft/client/gui/screens/Screen.java.patch (hunk 1 above)
Comment (Spinoscythe) [reply patches/net/minecraft/client/gui/screens/Screen.java.patch]: reverted!

File: patches/net/minecraft/client/gui/screens/inventory/AbstractContainerScreen.java.patch [hunk 4]
```diff
@@ -1,10 +1,12 @@
...
-@@ -92,6 +_,7 @@
+@@ -91,7 +_,8 @@
+     public void render(GuiGraphics p_283479_, int p_283661_, int p_281248_, float p_281886_) {
          int i = this.leftPos;
          int j = this.topPos;
-         super.render(p_283479_, p_283661_, p_281248_, p_281886_);
-+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ContainerScreenEvent.Render.Background(this, p_283479_, p_283661_, p_281248_));
+-        super.render(p_283479_, p_283661_, p_281248_, p_281886_);
++        // Neo: replace the super render call with renderBackground and manually render the widgets in the method below.
++        this.renderBackground(p_283479_, p_283661_, p_281248_, p_281886_);
```
Comment (XFactHD) [inline patches/net/minecraft/client/gui/screens/inventory/AbstractContainerScreen.java.patch:9]: The whole point of moving the implementation of the super method here is to not patch `renderBackground()` at all. Please move the event dispatch and the widget rendering loop here after the `renderBackground()` call and adjust the comment to `// Neo: replicate the super method's implementation to insert the event between background and widgets`

Comment (Spinoscythe) [reply patches/net/minecraft/client/gui/screens/inventory/AbstractContainerScreen.java.patch:9]: fixed!

**Review by XFactHD — DISMISSED** [review]

**Comment by XFactHD** [comment]
One last request: please add a comment to `Screen#render()` that mixins targeting it won't fire for container screens, i.e. something along the lines of `/** Neo: mixins targeting this method won't fire for container screens as {@link AbstractContainerScreen#render()} replicates this method in place of a super call to insert an event */`

**Review by Technici4n — APPROVED** [review]

**Pull request merged**